**NAMA: DIVA ANGGRAENI**

**NPM: 24083010065**

**MATEMATIKA DISKRIT (A)**

**PENUGASAN: Matriks Kovarian dan Korelasi serta Cosine Similarity**

**Selasa, 2 September 2025**

In [64]:
import pandas as pd
import numpy as np
df = pd.read_csv('/siswa.csv')

In [65]:
# Hilangkan tanda koma di kolom string
df['NAMA'] = df['NAMA'].str.replace(',', '')

# Simpan ke excel
df.to_excel("data_siswa.xlsx", index=False)

In [66]:
df = pd.read_excel('/data_siswa.xlsx')
df.head()

NPM                      NAMA  adil, amanah  suka membantu  \
0  24083010001     MADINA HEDY ALIFANTI            0.6            0.9   
1  24083010003      MARIA DWI KURNIASIH            0.6            1.0   
2  24083010005  ADINDA PUTRI RACHMAWATI            0.8            0.9   
3  24083010006           AQUINA SYABITA            1.0            0.9   
4  24083010007      VARID PUTRA PRATAMA            1.0            0.6   

   nyambung kalau ngobrol  asyik, menarik (D)  
0                     0.5                 0.7  
1                     0.7                 0.9  
2                     0.6                 0.5  
3                     0.9                 0.7  
4                     0.7                 0.8

**Tentukan prioritas/nilai beban untuk A, B, C, D sebagai matrix P; pilih empat orang (baris) dengan nilai terbesar dari matrix C.P**

In [67]:
C = df.iloc[:, 2:].to_numpy()
P = np.array([1.0, 0.8, 0.6, 0.9])

skor = C @ P
df['Skor'] = skor
df_sorted = df.sort_values(by='Skor', ascending=False)

# Top 4 mahasiswa dengan skor terbesar
top4 = df_sorted.head(4)

print("Empat mahasiswa dengan skor tertinggi:")
print(top4[['NAMA', 'Skor']])

Empat mahasiswa dengan skor tertinggi:
                                NAMA  Skor
38              KIKY MAUDRY NATASYA   2.98
3                    AQUINA SYABITA   2.89
8   RAVEENA AYU DESEMBER SURYOPUTRI   2.88
7           HIZKIA SAMHAN REZAYOSHI   2.80


Nilai beban (P) yang saya gunakan untuk setiap kriteria yaitu adil, amanah (1.0); suka membantu (0.8); nyambung kalau ngobrol (0.6); dan asyik, menarik (0.9). Empat mahasiswa dengan skor tertinggi ditentukan dari hasil penjumlahan matriks C.P dari setiap kriteria siswa. Dengan begitu berdasarkan total skor tersebut, diambil empat mahasiswa yang memiliki skor tertinggi untuk ditampilkan.

**Hitung matriks kovarian dan korelasi**
* cov_mat = np.cov(kelasku, rowvar=False, ddof=1)
* corr_mat = np.corrcoef(kelasku, rowvar=False)

In [68]:
kelasku = df.iloc[:, 2:]
cov_df = pd.DataFrame(np.cov(kelasku, rowvar=False, ddof=1), index=kelasku.columns, columns=kelasku.columns)
corr_df = pd.DataFrame(np.corrcoef(kelasku, rowvar=False), index=kelasku.columns, columns=kelasku.columns)

print("Matriks Kovarian")
display(cov_df)
print("Matriks Korelasi")
display(corr_df)

Matriks Kovarian


adil, amanah  suka membantu  nyambung kalau ngobrol  \
adil, amanah                0.031635      -0.003462               -0.002212   
suka membantu              -0.003462       0.027077               -0.001718   
nyambung kalau ngobrol     -0.002212      -0.001718                0.027122   
asyik, menarik (D)          0.000160      -0.004897                0.002006   
Skor                        0.027683       0.012762                0.014493   

                        asyik, menarik (D)      Skor  
adil, amanah                      0.000160  0.027683  
suka membantu                    -0.004897  0.012762  
nyambung kalau ngobrol            0.002006  0.014493  
asyik, menarik (D)                0.017173  0.012902  
Skor                              0.012902  0.058199

Matriks Korelasi


adil, amanah  suka membantu  nyambung kalau ngobrol  \
adil, amanah                1.000000      -0.118274               -0.075501   
suka membantu              -0.118274       1.000000               -0.063394   
nyambung kalau ngobrol     -0.075501      -0.063394                1.000000   
asyik, menarik (D)          0.006876      -0.227115                0.092969   
Skor                        0.645160       0.321473                0.364786   

                        asyik, menarik (D)      Skor  
adil, amanah                      0.006876  0.645160  
suka membantu                    -0.227115  0.321473  
nyambung kalau ngobrol            0.092969  0.364786  
asyik, menarik (D)                1.000000  0.408104  
Skor                              0.408104  1.000000

**Matriks kovarian:** hanya menunjukkan arah hubungan antar variabel (positif/negatif) sehingga tidak dapat diukur langsung seberapa kuat hubungan sebab  dipengaruhi skala data. Semua kriteria memiliki hubungan positif dengan skor yang berarti semakin tinggi nilai kriteria maka semakin tinggi juga skornya.

**Matriks korelasi:** memiliki nilai antara -1 sampai 1, sehingga lebih mudah diketahui seberapa kuat hubungan antar variabel. Dari hasil tersebut terlihat bahwa kriteria adil, amanah memiliki korelasi paling kuat dengan skor (0.65) yang berarti kriteria tersebut paling memengaruhi skor.

**Hitung cosine similarity antar siswa**
* norms = np.linalg.norm(X_matrix, axis=1, keepdims=True)
* cosine_sim = (X_matrix @ X_matrix.T) / (norms @ norms.T)
* cosine_df = pd.DataFrame(cosine_sim, index=students, columns=students)


In [69]:
matrix = df.iloc[:, 2:6].to_numpy()

norms = np.linalg.norm(matrix, axis=1, keepdims=True)
cosine_sim = (matrix @ matrix.T) / (norms @ norms.T)
cosine_df = pd.DataFrame(cosine_sim, index=df['NAMA'], columns=df['NAMA'])

print("Cosine Similarity Antar Siswa")
cosine_df

Cosine Similarity Antar Siswa


NAMA                                    MADINA HEDY ALIFANTI   \
NAMA                                                            
MADINA HEDY ALIFANTI                                 1.000000   
MARIA DWI KURNIASIH                                  0.993781   
ADINDA PUTRI RACHMAWATI                              0.978028   
AQUINA SYABITA                                       0.964209   
VARID PUTRA PRATAMA                                  0.940021   
GAITSA NAZWA KANSA                                   0.979694   
Mohammad Alimun Hakim                                0.948556   
HIZKIA SAMHAN REZAYOSHI                              0.990850   
RAVEENA AYU DESEMBER SURYOPUTRI                      0.974599   
Yohanes Olvin Jun Sole                               0.979058   
AISYAH AMALIA HAMID                                  0.916230   
FEBRIANI YOLANDA TESSALONIKA                         0.914364   
AMELIA RIZQYNA PUTRI                                 0.935773   
VIA AMANDA                                           0.969083   
FANNY WIDYA CAHYANI                                  0.965481   
SITI NAIA HESTI RACHMAWATI                           0.962154   
ALFANI NUR AZIZAH                                    0.933717   
Diah Anggraini                                       0.957309   
AHMAD KENZY FARZAQ                                   0.987822   
Angelina Nirmala Puteri Dika Praktiko                0.956197   
MAULIDA APRILIA PUTRI HANDAYANI                      0.926285   
Izzati Kamila Putri                                  0.978028   
AULIYA KHOTIMATUZ ZAHROH                             0.931937   
CARISSA NAURA RAJWA                                  0.974805   
DIVA ANGGRAENI                                       0.984047   
ERIK SAPUTRA RIFKI                                   0.982745   
MAYA PURNAMA SARI                                    0.982600   
ACHMAD DANY GUNAWAN                                  0.982745   
Gendis Poerbodani                                    0.984224   
MUHAMMAD WILDAN SULTANSYAH                           0.989640   
ESTHI NURANI SRI HANDAYANI                           0.933357   
Muhammad Naufal Asruzzaman                           0.959041   
Laudya Meitaneia Sianturi                            0.987664   
ADRIAN VEDA DARMAWAN                                 0.939565   
SELVY DWI YULITA SARI                                0.927085   
Rizky Faza Hafiyyan Nusantara                        0.993333   
Indra Maulana R F Y                                  0.965576   
YUNIAR RACHMAWATI                                    0.978774   
KIKY MAUDRY NATASYA                                  0.969817   
NAJWA SOFIA                                          0.906240   

NAMA                                    MARIA DWI KURNIASIH   \
NAMA                                                           
MADINA HEDY ALIFANTI                                0.993781   
MARIA DWI KURNIASIH                                 1.000000   
ADINDA PUTRI RACHMAWATI                             0.961187   
AQUINA SYABITA                                      0.959595   
VARID PUTRA PRATAMA                                 0.936433   
GAITSA NAZWA KANSA                                  0.981834   
Mohammad Alimun Hakim                               0.973655   
HIZKIA SAMHAN REZAYOSHI                             0.990323   
RAVEENA AYU DESEMBER SURYOPUTRI                     0.976325   
Yohanes Olvin Jun Sole                              0.976035   
AISYAH AMALIA HAMID                                 0.940542   
FEBRIANI YOLANDA TESSALONIKA                        0.913581   
AMELIA RIZQYNA PUTRI                                0.934550   
VIA AMANDA                                          0.973095   
FANNY WIDYA CAHYANI                                 0.963656   
SITI NAIA HESTI RACHMAWATI                          0.969069   
ALFANI NUR AZIZAH                                   0.947758   
Diah Anggraini                                      0

Pada cosine similarity ini mengukur kemiripan antar siswa berdasarkan nilai kriterianya. Contohnya pada baris pertama, Madina dengan dirinya sendiri nilainya 1.000000 pasti sama persis. Jika Madina dibandingkan Najwa Sofia, tingkat kemiripan kriterianya 0.906240 masih mirip namun masih lebih rendah dibandingkan Madina dengan Maria yang nilainya 0.993781 (hampir mirip).

**Analisis tambahan (clustering)**

In [70]:
from sklearn.cluster import KMeans
import numpy as np

X = df[["Skor"]].to_numpy()
kmeans = KMeans(n_clusters=2, random_state=42, n_init=10)
df["Cluster"] = kmeans.fit_predict(X)
print("Cluster mahasiswa berdasarkan skor:")
print(df[["NAMA", "Skor", "Cluster"]].sort_values("Skor", ascending=False))

# Keterangan rata-rata skor tiap cluster
print("\nRata-rata skor per cluster:")
print(df.groupby("Cluster")["Skor"].mean())

Cluster mahasiswa berdasarkan skor:
                                      NAMA  Skor  Cluster
38                    KIKY MAUDRY NATASYA   2.98        1
3                          AQUINA SYABITA   2.89        1
8         RAVEENA AYU DESEMBER SURYOPUTRI   2.88        1
7                 HIZKIA SAMHAN REZAYOSHI   2.80        1
12                   AMELIA RIZQYNA PUTRI   2.78        1
15             SITI NAIA HESTI RACHMAWATI   2.75        1
27                    ACHMAD DANY GUNAWAN   2.67        1
25                     ERIK SAPUTRA RIFKI   2.67        1
1                     MARIA DWI KURNIASIH   2.63        1
4                     VARID PUTRA PRATAMA   2.62        1
14                    FANNY WIDYA CAHYANI   2.60        1
20        MAULIDA APRILIA PUTRI HANDAYANI   2.60        1
33                   ADRIAN VEDA DARMAWAN   2.59        1
36                    Indra Maulana R F Y   2.59        1
19  Angelina Nirmala Puteri Dika Praktiko   2.57        1
11           FEBRIANI YOLANDA TESSAL

Clustering ini mengelompokkan mahasiswa berdasarkan skor C.P yang sudah dihitung sebelumnya. Cluster 0 memiliki rata-rata skor 2.2860, lebih rendah dibandingkan cluster 1 yang memiliki rata-rata skor 2.6625. Mahasiswa dengan cluster 1 dianggap sebagai ketua kelompok sebab berdasarkan skor penilaian kriteria ia lebih unggul dibandingkan yang lain, sehingga layak menjadi pemimpin. Sedangkan, mahasiswa dengan cluster 0 dianggap sebagai anggota kelompok sebab skor penilaian kriteria lebih rendah dibandingkan dengan cluster 1. Dengan begitu, clustering ini membantu dalam pengelompokan mahasiswa secara objektif berdasarkan skor penilaian karakter yang diperoleh.